cpp/src/arrow/array/validate.cc

```diff
@@ -558,7 +559,7 @@ struct ValidateArrayImpl {
     }
 
     if (full_validation) {
-      if (data.null_count != kUnknownNullCount) {
+      if (data.null_count != kUnknownNullCount || !nullable) {
```

Comment:
Ok, but where does it test that `actual_null_count` is 0 if `nullable` is false?